### PR TITLE
Enable dotnet test in libraries

### LINKED ISSUE #945: Enable dotnet test in dotnet/runtime
Investigate how much work it is to switch away from the deprecated xunit.console runner to dotnet vstest. This is just about the runner, the test infra backend will still be xunit.

UAP will continue using Microsoft.DotNet.XUnitRunnerUap.
UAPAOT is problematic. We currently use xunit.console for it which doesn't work (https://github.com/dotnet/corefx/issues/30423). It may be our best option to add a ifdef configuration in   https://github.com/dotnet/arcade/pull/1627 which works fine with uapaot.

### PR DESCRIPTION
**Pull request by ViktorHofer**
**No Merge. Part of batched rollout in May: https://github.com/dotnet/runtime/issues/35202**

Fixes https://github.com/dotnet/runtime/issues/945
Requires https://github.com/dotnet/runtime/pull/35538

- Stop inheriting `IsTestSupportProject` from `IsTestProject` as the VSTest target runs based on the latter one. This enables dotnet test on sln projects.
- Upgrading the SDK for vstest default filter support in .runsettings files.
- Replacing the coverlet.console with the inbuilt VSTest coverlet datacollector.
- Add both TRX and HTML logger for test runs
- Remove dead code and fix chmod condition
- Remove obsolete vstest.props/targets files as VSTest can now be invoked directly via `dotnet test`

### Feature additions
- All of dotnet test can be used (either on a project or directly on the compiled assembly) means advanced filtering, test-listing (without running), blame data-collector, html logger, etc.
- HTML logger
- Ability to run dotnet test directly in the repo root (e.g. `dotnet test -f netcoreapp5.0 /p:Subset=libs.tests` or on a solution (e.g. `dotnet test -f netcoreapp5.0 src\libraries\System.Text.RegularExpressions`)
- Enabling F5 support for both .NETCoreApp and .NETFramework (+ without the need for launchSettings.json files).

### Behavior Change
For measuring code coverage, the inbuilt dotnet test data collector is now used (which improves perf and stability and is actually the only supported way to measure coverage with dotnet test). The command changes from `dotnet build /t:Test /p:Coverage=true` to `dotnet test -f netcoreapp5.0 /p:Coverage=true`.

### Plan for migrating from xunit.console -> dotnet test
1. This PR adds the option to use `dotnet test`.
2. In the (hopefully near) future when we believe that we covered all existing use cases with `dotnet test` without regressing the existing workflow and efficiency, we will add a warning to the test target indicating that it will be removed with the next batched rollout and will recommend to switch to `dotnet test`.
3. In the then next batched rollout, the Test target will be removed.

cc @stephentoub @danmosemsft

### CODE REVIEW

**Comment by ViktorHofer** [comment]
cc @davidfowl

**Comment by stephentoub** [comment]
> Part of batched rollout in May

What is the impact this has on the commands we use in the repo?

File: docs/workflow/building/libraries/code-coverage.md [hunk 1]
```diff
@@ -28,40 +28,30 @@ An issue need not be addressed in its entirety. We happily accept contributions
 
 You can perform code coverage runs for the entire repository locally by adding the `coverage` switch (assuming that  source and test assemblies are already built):
 
-    build -test -coverage
+    build libs -test -coverage
```
Comment (safern) [inline docs/workflow/building/libraries/code-coverage.md:31]: the default libs subset doesn't build the tests right?

File: eng/testing/tests.props [hunk 2]
```diff
@@ -5,6 +5,19 @@
     <!-- Implicit test build support when invoking the Test target on the project directly. -->
     <TestDependsOn Condition="'$(TestNoBuild)' != 'true' and '$(BuildAllProjects)' != 'true'">Build</TestDependsOn>
     <TestDependsOn>$(TestDependsOn);GenerateRunScript;RunTests</TestDependsOn>
+    <VSTestNoLogo>true</VSTestNoLogo>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <!-- Default and user defined categories -->
```
Comment (safern) [inline eng/testing/tests.props:12]: Why is this needed here?

File: src/libraries/Directory.Build.props [hunk 3]
```diff
@@ -322,35 +321,34 @@
...
-  <ItemGroup Condition="'$(IsTestProject)' == 'true' and '$(IsTestSupportProject)' != 'true'">
+  <ItemGroup Condition="'$(IsTestProject)' == 'true'">
     <ProjectReference Include="$(CommonTestPath)TestUtilities\TestUtilities.csproj" />
   </ItemGroup>
 
-  <PropertyGroup Condition="'$(IsTestProject)' == 'true' and '$(IsTestSupportProject)' != 'true'">
+  <PropertyGroup Condition="'$(IsTestProject)' == 'true'">
     <EnableTestSupport>true</EnableTestSupport>
-    <EnableCoverageSupport Condition="'$(Coverage)' == 'true' and '$(CoverageSupported)' != 'false'">true</EnableCoverageSupport>
+    <EnableCoverageSupport>true</EnableCoverageSupport>
```
Comment (safern) [inline src/libraries/Directory.Build.props:334]: Why are we enabling test coverage always?

File: docs/workflow/building/libraries/code-coverage.md (hunk 1 above)
Comment (ViktorHofer) [reply docs/workflow/building/libraries/code-coverage.md:31]: Thanks :)

File: eng/testing/tests.props (hunk 2 above)
Comment (ViktorHofer) [reply eng/testing/tests.props:12]: This was moved from xunit.targets as it isn't xunit specific and runsettings (which is runner agnostic) needs access to it.

File: src/libraries/Directory.Build.props (hunk 3 above)
Comment (ViktorHofer) [reply src/libraries/Directory.Build.props:334]: This is just the infra support. We import it for all test projects and for tests.proj.

File: eng/testing/tests.props (hunk 2 above)
Comment (safern) [reply eng/testing/tests.props:12]: Cool.

File: src/libraries/Directory.Build.props (hunk 3 above)
Comment (safern) [reply src/libraries/Directory.Build.props:334]: Sounds good.

**Comment by ViktorHofer** [comment]
> What is the impact this has on the commands we use in the repo?

Sorry, mentioned the behavior change only in the batched rollout issues and forgot to add it here as well. Please see the updated top post.

**Review by akoeplinger — APPROVED** [review]

**Comment by EgorBo** [comment]
@ViktorHofer 
um.. so it broke running tests for Android and iOS 😢 
Any idea how can I run them now?
I used to run them via
```
./build.sh -os Android -arch arm64 -test
```
Now it's no-op.

Running individual tests still works:
```
cd src/libraries/System.Collections/tests
dotnet build /t:Test /p:TargetOS=Android /p:TargetArchitecture=arm64
```
But I don't know which folder and how should I run this command from

**Comment by ViktorHofer** [comment]
> Now it's no-op.

It's not doing anything?

**Comment by ViktorHofer** [comment]
Just tried this out locally and the test execution of all projects still work. Can I try your command out somehow?

**Comment by EgorBo** [comment]
> It's not doing anything?

@ViktorHofer it only builds libs but doesn't launch `RunTests.sh`

> Can I try your command out somehow?

Oh, you will need to install Android NDK/SDK so I better try to figure out myself. So are you saying `-test` should still work?

**Comment by ViktorHofer** [comment]
Isn't the command build.sh libs.tests - test? I don't see the subset specified in your command.

**Comment by EgorBo** [comment]
@ViktorHofer 
I've rebuilt the repo and try again and now it works so sorry, everything is ok 👍

**Comment by ViktorHofer** [comment]
No problem at all :)
